src/types/list.rs

```diff
@@ -18,30 +20,79 @@ pub struct PyList(PyAny);
-    for (i, e) in elements.enumerate() {
-        let obj = convert(e).into_ptr();
-        #[cfg(not(Py_LIMITED_API))]
-        ffi::PyList_SET_ITEM(ptr, i as Py_ssize_t, obj);
-        #[cfg(Py_LIMITED_API)]
-        ffi::PyList_SetItem(ptr, i as Py_ssize_t, obj);
-    }
-    ptr
+    unsafe {
+        // PyList_New checks for overflow but has a bad error message, so we check ourselves
+        let len: Py_ssize_t = elements.len().try_into().unwrap_or_else(|_| {
+            panic!("out of range integral type conversion attempted on `elements.len()`")
+        });
+
+        let list = ffi::PyList_New(len);
+
+        // - Panics if the ptr is null
+        // - Cleans up the list if `convert` or the asserts panic
+        let guard: Py<PyAny> = Py::from_owned_ptr(py, list);
+
+        let mut counter: Py_ssize_t = 0;
+
+        for (i, e) in elements.by_ref().enumerate().take(len as usize) {
+            let obj = convert(e).into_ptr();
+            #[cfg(not(Py_LIMITED_API))]
+            ffi::PyList_SET_ITEM(list, i as Py_ssize_t, obj);
+            #[cfg(Py_LIMITED_API)]
+            ffi::PyList_SetItem(list, i as Py_ssize_t, obj);
+            counter += 1;
+        }
+
+        assert!(elements.next().is_none(), "Attempted to create PyList but `elements` was larger than reported by its `ExactSizeIterator` implementation.");
+        assert_eq!(len, counter, "Attempted to create PyList but `elements` was smaller than reported by its `ExactSizeIterator` implementation.");
+
+        std::mem::forget(guard);
+        list
+    }
 }
 
 impl PyList {
     /// Constructs a new list with the given elements.
+    ///
+    /// If you want to create a [`PyList`] with elements of different or unknown types, or from an
+    /// iterable that doesn't implement [`ExactSizeIterator`], use [`PyList::append`].
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use pyo3::prelude::*;
+    /// use pyo3::types::PyList;
+    ///
+    /// # fn main() {
+    /// Python::with_gil(|py| {
+    ///     let elements: Vec<i32> = vec![0, 1, 2, 3, 4, 5];
+    ///     let list: &PyList = PyList::new(py, elements);
+    ///     assert_eq!(format!("{:?}", list), "[0, 1, 2, 3, 4, 5]");
+    /// });
+    /// # }
+    /// ```
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if `element`'s [`ExactSizeIterator`] implementation is incorrect.
+    /// All standard library structures implement this trait correctly, if they do, so calling this
+    /// function with (for example) [`Vec`]`<T>` or `&[T]` will always succeed.
+    #[track_caller]
     pub fn new<T, U>(py: Python<'_>, elements: impl IntoIterator<Item = T, IntoIter = U>) -> &PyList
     where
         T: ToPyObject,
         U: ExactSizeIterator<Item = T>,
     {
         unsafe {
-            py.from_owned_ptr::<PyList>(new_from_iter(elements.into_iter(), |e| e.to_object(py)))
+            py.from_owned_ptr::<PyList>(new_from_iter(py, elements.into_iter(), |e| {
```

Comment:
How about `.map()` instead of making a separate `convert` argument?

I also wonder if we should consider using `&mut dyn ExactSizeIterator<Item = PyObject>` as the argument to `new_from_iter`, to avoid generic code bloat?